File: daemon/src/ws/handlers/debugLog.ts
import type { WSConnection } from "../connection.js";
import type { ClientMessage } from "../protocol.js";
import { appendDebug } from "../../debugLog.js";

/**
 * Persist client-shipped input/composition events (mobile double-text
 * investigation). Receiving any debug:log marks the connection debug-active so
 * its session:input bytes are logged too (see sessionInput.ts).
 */
export function handleDebugLog(
  conn: WSConnection,
  msg: Extract<ClientMessage, { type: "debug:log" }>,
): void {
  conn.debugInput = true;
  for (const entry of msg.entries) {
    appendDebug({ src: "client", conn: conn.id, ...entry });
  }
}

File: rust/vst-ws/src/handlers/debug_log.rs
//! `debug:log` handler — marks the connection debug-active and persists
//! client-shipped input/composition events (mobile double-text investigation).

use vst_types::ws::ClientMessage;

use crate::connection::WsConnection;

pub fn handle_debug_log(conn: &WsConnection, msg: &ClientMessage) {
    if let ClientMessage::DebugLog { entries } = msg {
        conn.set_debug_input(true);
        // Entries are diagnostic-only; the TS `appendDebug` persists them to a
        // log. Here we simply record that the connection is debug-active and
        // accept the entries (wire-validated).
        let _ = entries;
    }
}
